File: leecode/401-500/468-验证IP地址.java
class Solution {
    public String validIPAddress(String IP) {
        if (IP == null) {
            return "Neither";
        }
        
        String regex0 = "(\\d|[1-9]\\d|1\\d\\d|2[0-4]\\d|25[0-5])";
        String regexIPv4 = regex0 + "(\\." + regex0 + "){3}";
        String regex1 = "([\\da-fA-F]{1,4})";
        String regexIPv6 = regex1 + "(:" + regex1 + "){7}";
        
        String result = "Neither";
        if (IP.matches(regexIPv4)) {
            result = "IPv4";
        } else if (IP.matches(regexIPv6)) {
            result = "IPv6";
        }
        return result;
    }

    public static void main(String[] args) {
        Solution a = new Solution();
        String x = a.validIPAddress("2001:0db8:85a3:0:0:8A2E:0370:7334");
        System.out.println(x);
    }
}

File: leecode/401-500/468-验证IP地址.py
import re

def validIPAddress(IP):
    if not IP:
        return "Neither"
    
    regex0 = r"\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]"
    regexIPv4 = regex0 + "(\\." + regex0 + "){3}"
    regex1 = r"[\da-fA-F]{1,4}"
    regexIPv6 = regex1 + "(:" + regex1 + "){7}"
    result = "Neither"
    if re.match(regexIPv4, IP):
        result = "IPv4"
    elif re.match(regexIPv6, IP):
        a = re.match(regexIPv6, IP)
        print(a.group(0))
        print(len(a.groups()))
        
        result = "IPv6"
    
    
    return result

a = validIPAddress("172.16.254.1")
print(a)

